crates/ra_hir/src/debug.rs

```diff
@@ -0,0 +1,84 @@
+//! printf debugging infrastructure for rust-analyzer.
+//!
+//! When you print a hir type, like a module, using `eprintln!("{:?}", module)`,
+//! you usually get back a numeric ID, which doesn't tell you much:
+//! `Module(92)`.
+//!
+//! This module adds convenience `debug` methods to various types, which resolve
+//! the id to a human-readable location info:
+//!
+//! ```not_rust
+//! eprintln!("{:?}", module.debug(db));
+//! =>
+//! Module { name: collections, path: "liballoc/collections/mod.rs" }
+//! ```
+//!
+//! Note that to get this info, we might need to execute queries! So
+//!
+//! * don't use the `debug` methods for logging
+//! * when debugging, be aware that interference is possible.
+
+use std::{cell::Cell, fmt};
+
+use ra_db::{CrateId, FileId};
+
+use crate::{db::HirDatabase, Crate, Module, Name};
+
+impl Crate {
+    pub fn debug(self, db: &impl HirDebugDatabase) -> impl fmt::Debug + '_ {
```

Comment:
why not start out making this a trait? :thinking:

Reply:
inherent method is more convenient for debugging as you don't have to add an import, and we don't have to generalize over this functionality yet